moquette/broker/src/main/java/io/moquette/broker/Session.java

```diff
@@ -113,6 +113,34 @@ public int compareTo(Delayed o) {
         this.sessionQueue = sessionQueue;
     }
 
+    /**
+     * This method terminates a Session, which involves disconnecting the client,
+     * if one is connected, and releasing all references to unacknowledged publishes.
+     */
+    void terminateSession() {
+        MQTTConnection connection = this.mqttConnection;
+        if (connection != null) {
+            connection.dropConnection();
+        }
+        disconnect();
+
+        // Release queued messages
+        while (!sessionQueue.isEmpty()) {
+            final SessionRegistry.EnqueuedMessage msg = sessionQueue.remove();
+            msg.release();
+        }
+
+        // Release in flight messages (QoS 1/2)
```

Comment:
TODO: this can possibly still throw a concurrent modification exception if another message gets delivered while this session is terminating, which could cause connect failures (and the leak of whatever is left in the inflight list)

Reply:
is it `inflightWindow.remove` that throws the modification exception?

Reply:
Yes. My previous PR that introduces reference counting was meant to address that and prevent concurrent modifications. I'd like to also pull that back in, but I think this is a good improvement as it stands, even if not perfect